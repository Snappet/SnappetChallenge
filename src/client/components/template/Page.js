import React from 'react';
import t from 'prop-types';
import styled from 'styled-components';
import {
  Drawer, List, ListItem, ListItemIcon as MuiListItemIcon, ListItemText as MuiListItemText
} from '@material-ui/core';
import {Assignment as ReportIcon, InfoRounded as InfoIcon} from '@material-ui/icons';
import {useDispatch, useSelector} from 'react-redux';
import {identity, view} from 'ramda';
import {createSelector} from 'reselect';
import {Link} from 'react-router-dom';

import View from '../atoms/View';
import AppBar from '../organisms/AppBar';
import theme from '../../theme';
import {drawerIsOpenState} from '../../redux/reducers/ui';
import {hideDrawer} from '../../redux/actions/ui';

const Content = styled(View)`
  padding: ${theme.spacing(2)}px;
  justify-content: stretch;
  align-items: stretch;
`;

const ListItemIcon = styled(MuiListItemIcon)`
  min-width: 36px;
`;

const ListItemText = styled(MuiListItemText)`
  a {
    text-decoration: none;
  }
`;

export const Page = ({title, drawerIsOpen, onCloseDrawer, children}) => (
  <View flex="1" alignItems="stretch">
    <Drawer anchor="left" open={drawerIsOpen} onClose={onCloseDrawer}>
      <List>
        <ListItem>
          <ListItemIcon><ReportIcon /></ListItemIcon>
          <ListItemText>
            <Link to="/">Report</Link>
          </ListItemText>
        </ListItem>
        <ListItem>
          <ListItemIcon><InfoIcon /></ListItemIcon>
          <ListItemText>
            <Link to="/about">About</Link>
          </ListItemText>
        </ListItem>
      </List>
    </Drawer>
    {title && <AppBar title={title} />}
    <Content>
      {children}
    </Content>
  </View>
);

Page.propTypes = {
  title: t.oneOfType([t.string, t.element]),
  children: t.oneOfType([
    t.element, t.arrayOf(t.element)
  ]).isRequired,
  drawerIsOpen: t.bool,
  onCloseDrawer: t.func,
};

const isOpenSelector = createSelector([view(drawerIsOpenState)], identity);

const ConnectedPage = args => {
  const drawerIsOpen = useSelector(isOpenSelector);
  const dispatch = useDispatch();

  const onCloseDrawer = () => dispatch(hideDrawer());

  return <Page onCloseDrawer={onCloseDrawer} drawerIsOpen={drawerIsOpen} {...args} />;
};

export default ConnectedPage;
