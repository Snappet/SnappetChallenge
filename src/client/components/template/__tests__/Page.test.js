import React from 'react';
import renderer from 'react-test-renderer';

import {Page} from '../Page';

describe('components', () => {
  describe('templates', () => {
    describe('Page', () => {
      test('should render Page with the provided title and children', () => {
        const $element = renderer.create(
          <Page>
            <h3>An random h1 title</h3>
          </Page>
        );
        const tree = $element.toJSON();
        const root = $element.root;

        expect(tree).toMatchSnapshot();
        expect(root.findByType('h3').children).toEqual(['An random h1 title']);
      });
    });
  });
});
