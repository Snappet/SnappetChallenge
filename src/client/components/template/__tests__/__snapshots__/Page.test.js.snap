// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components templates Page should render Page with the provided title and children 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: inherit;
  -ms-flex: inherit;
  flex: inherit;
}

.c2 {
  padding: 16px;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  >
    <h3>
      An random h1 title
    </h3>
  </div>
</div>
`;
